File: backend/schemas/flow.py
"""What a flow shows: boxes, and the arrows that say one leads to another.

Its own module for the reason the chart, the gantt and the mesh have one — a
payload made of two further models is not one more block of fields, and
``schemas.payloads`` is at the house limit of 350 lines. ``hud_mcp/flows.py`` is
the tool that matches it.

The geometry here is copied inward from ``scorsese_core::shape``, which solved
the same problem for a video renderer: a diagram that has to survive being
re-rendered at a different size. Every number is a **fraction of the widget** —
``x`` and ``w`` of its width, ``y`` and ``h`` of its height — so the drawing
means the same thing in a 3x2 widget and in a 16x9 one. Nothing here is a pixel
and nothing here is stored computed: where a node without a position ends up is
a reading the browser takes, the way ``GanttPayload`` refuses to store its own
window.

**The word `node` is correct here and only here.** ``CLAUDE.md`` bans it for
widgets because a node suggests a graph; inside a flow there is a graph. The
``text-node-*`` type scale in the stylesheet is the older, unrelated sense of
the word and means *widget*.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from schemas.colour import Colour
from schemas.icon import Icon

# The two shapes a box comes in, and the list is meant to stay this short.
# scorsese holds the same line and gives the reason: a rectangle, an ellipse and
# an arrow are what a diagram is made of — boxes are its nouns and arrows its
# verbs, and polygons, stars and arbitrary paths are a drawing program growing
# inside something that was not one. A decision is said by the words on the two
# arrows leaving a node, not by a diamond nobody can read from a sofa.
#
# `ellipse` rather than `circle` because the two dimensions are fractions of
# different things — width of the widget's width, height of its height — so
# picking one axis to measure both against would be a different bug in every
# widget size.
FlowShape = Literal["rectangle", "ellipse"]

# Where on a box an arrow meets it, as ``scorsese_core::shape::Side``.
#
# `center` is the honest answer when an arrow is meant to point *at* a box
# rather than to touch it; it is drawn over the box, which is what was asked
# for.
FlowSide = Literal["left", "right", "top", "bottom", "center"]

# How an arrow gets from one end to the other. `s` leaves each end along that
# end's own outward normal, which is the shape two boxes side by side want — a
# straight diagonal between them reads as a mistake. Which axis it bows along is
# deliberately not a field; scorsese: "it is not a question an author has an
# opinion about."
FlowCurve = Literal["straight", "s"]

# Which ends of an arrow carry a head. This is what makes a plain connecting
# line not a kind of its own — scorsese again: "it is an arrow that points at
# nothing." `end` is the default because an arrow is drawn to say *this leads to
# that*, and that reading has a direction.
FlowHeads = Literal["none", "end", "both"]

# The most a corner can be rounded: half the box's shorter side, where the two
# corners of that side meet and the end has become a semicircle. Past it there
# is no straight edge left to round.
MAX_RADIUS = 0.5


class FlowNode(BaseModel):
    """One box in a flow, with a word in it.

    A box either names its whole rectangle — all four of ``x``, ``y``, ``w``,
    ``h`` — or names none of it and is laid out for. Half a rectangle has no
    honest reading, and "the widget will fit the rest around what you gave it"
    is a layout engine with a special case in it.

    Nodes may overlap each other and nothing enforces distance between them. The
    board refuses overlap because the board auto-places and is finite; inside a
    flow the author placed both deliberately, and a small box drawn over a larger
    one is a legitimate drawing.
    """

    model_config = ConfigDict(extra="forbid")

    # What a link names this box by. Unique within the flow — a duplicate is
    # refused naming it, because a link to the second of two boxes called
    # "build" would silently draw to the first.
    id: str
    text: str
    shape: FlowShape = "rectangle"
    # How rounded the corners are, as a fraction of the box's **own shorter
    # side** rather than of the widget: 0 is a square corner and 0.5 is a pill,
    # whatever size the box is and whatever shape the widget is. Copied outright
    # from ``Geometry::Rectangle``, whose comment gives the reason — a widget
    # fraction could be larger than the box it rounds, which is nonsense nothing
    # can catch until something works out the pixels. It also keeps corners
    # circular instead of elliptical, since one number then means one distance
    # rather than two. Ignored by an ellipse, which is all corner.
    radius: float = Field(default=0.18, ge=0.0, le=MAX_RADIUS)
    # The box's colour: its interior is a wash of this and its outline is this
    # at full strength — *wash areas, never marks*. Left out it takes the
    # widget's own ink, which over the board's video is a pane of smoked glass.
    # An eight-digit hex carries its own alpha and is drawn as written.
    color: Colour | None = None
    # Where the box sits and how big it is, all in 0-1 from the widget's
    # top-left. All four or none. Out of 0-1 is refused rather than clipped,
    # this widget's far edge included: it does not scroll and there is no
    # off-screen for anything to come in from.
    x: float | None = None
    y: float | None = None
    w: float | None = None
    h: float | None = None

    @model_validator(mode="after")
    def _a_whole_rectangle_or_none_of_one(self) -> "FlowNode":
        """Refuse half a box, and one that would be drawn off the widget."""
        given = [n for n, v in self._rect().items() if v is not None]
        if given and len(given) < 4:
            raise ValueError(
                f"node {self.id!r} gives only {', '.join(given)}; a box is placed by all "
                f"four of x, y, w and h, or by none of them"
            )
        if not given:
            return self
        for edge, near, far in (("x", self.x, self.w), ("y", self.y, self.h)):
            assert near is not None and far is not None
            if far <= 0:
                raise ValueError(f"node {self.id!r} has no width or height to draw")
            if near < 0 or near + far > 1:
                raise ValueError(
                    f"node {self.id!r} runs from {edge}={near:g} to {near + far:g}, which "
                    f"is outside the widget; a flow is clipped, not scrolled, so every "
                    f"edge has to land inside 0-1"
                )
        return self

    def _rect(self) -> dict[str, float | None]:
        """The four numbers that place this box, by name."""
        return {"x": self.x, "y": self.y, "w": self.w, "h": self.h}

    @property
    def placed(self) -> bool:
        """Whether this box says where it sits."""
        return self.x is not None


class FlowLink(BaseModel):
    """An arrow from one box to another.

    ``source`` and ``target`` rather than ``from`` and ``to``: ``from`` is a
    Python keyword, and the usual answer — ``from_`` with an alias — is a trap
    in this codebase specifically. The board is serialised down three different
    paths, and only one of them passes ``by_alias``, so the same link would
    reach the browser as ``from`` over HTTP and ``from_`` over the socket. One
    spelling everywhere is worth more than the nicer word.
    """

    model_config = ConfigDict(extra="forbid")

    source: str
    target: str
    # Which side of each box the arrow meets, or None to let the widget pick the
    # pair of facing sides. This is a deliberate divergence from scorsese, which
    # *requires* the author to choose because there an attached clip moves over
    # time and an arrow picking its own side would rearrange itself between two
    # renders. Nothing in a flow moves: the layout is a pure function of the
    # payload, so the same payload draws the same arrow forever.
    source_side: FlowSide | None = None
    target_side: FlowSide | None = None
    # A word on the arrow — what makes two arrows out of one box a decision
    # rather than a fork. Dropped when the arrow is too short to hold it: a word
    # half-overlapping a line is worse than no word.
    label: str | None = None
    curve: FlowCurve = "straight"
    heads: FlowHeads = "end"
    # Full-strength ink, whatever it is: an arrow is a mark and marks are never
    # washed. A hairline at a third strength over a moving video is not faint
    # from six feet away, it is absent.
    color: Colour | None = None


class FlowPayload(BaseModel):
    """A diagram of boxes and arrows, drawn whole inside one widget.

    The thing no other widget here can say: *this leads to that*. A deployment,
    a morning routine, the shape of a pipeline, a decision with two ways out.

    **The diagram is the widget and the arrow is a line inside it.** A
    free-standing arrow joining two widgets would make the board a canvas, every
    widget a node and dragging load-bearing — the exact opposite of what
    ``CLAUDE.md`` settles under *looks beat handling*. So a flow is drawn whole,
    the way a chart draws its own axes.

    Nothing in it is interactive: no click, no hover, no selection, no per-node
    handle. The television has no pointer and the sofa has no keyboard.

    A flow that changes is rewritten whole by ``key`` through the panel path.
    There is deliberately no ``add_to_flow`` and no per-node write until a writer
    exists that genuinely cannot rewrite the whole thing.
    """

    # Set here rather than inherited, like the chart, the gantt and the mesh:
    # this module cannot import the base in ``payloads`` without a cycle, since
    # that module imports this one.
    model_config = ConfigDict(extra="forbid")

    kind: Literal["flow"] = "flow"
    title: str | None = None
    icon: Icon | None = None
    nodes: list[FlowNode] = []
    links: list[FlowLink] = []

    @model_validator(mode="after")
    def _a_diagram_that_can_be_drawn(self) -> "FlowPayload":
        """Refuse the four ways a flow is wrong, each in a sentence naming it.

        Refused rather than patched up, for the reason the radial refuses a
        fourth ring: a widget that silently draws something other than what was
        dictated is invisible from the sofa and the caller is never told. The
        sentence is the whole product here — whoever wrote this cannot see the
        board and has only this line to work out what it typed wrongly.
        """
        self._one_box_per_name()
        self._all_placed_or_none_placed()
        self._every_arrow_joins_two_boxes()
        return self

    def _one_box_per_name(self) -> None:
        """A link names a box; two boxes with one name make that ambiguous."""
        seen: set[str] = set()
        for node in self.nodes:
            if node.id in seen:
                raise ValueError(
                    f"two nodes are called {node.id!r}; a link names a node by its id, so "
                    f"the ids have to be different"
                )
            seen.add(node.id)

    def _all_placed_or_none_placed(self) -> None:
        """Either every box says where it sits, or none does."""
        placed = [n.id for n in self.nodes if n.placed]
        loose = [n.id for n in self.nodes if not n.placed]
        if placed and loose:
            raise ValueError(
                f"{', '.join(repr(n) for n in placed)} say where they sit and "
                f"{', '.join(repr(n) for n in loose)} do not; place every node or none of "
                f"them, because a half-placed flow has no honest reading"
            )

    def _every_arrow_joins_two_boxes(self) -> None:
        """Both ends of every arrow name a box, and two different ones."""
        known = {n.id for n in self.nodes}
        drawn: set[tuple[str, str]] = set()
        for link in self.links:
            for end, name in (("source", link.source), ("target", link.target)):
                if name not in known:
                    raise ValueError(
                        f"a link's {end} is {name!r}, which is not a node in this flow; "
                        f"it has {', '.join(sorted(repr(n) for n in known)) or 'no nodes'}"
                    )
            if link.source == link.target:
                raise ValueError(
                    f"a link joins {link.source!r} to itself; a flow draws no loop back "
                    f"into the same node"
                )
            pair = (link.source, link.target)
            if pair in drawn:
                raise ValueError(
                    f"{link.source!r} leads to {link.target!r} twice; a flow draws one "
                    f"arrow per pair, so the second would be laid exactly over the first"
                )
            drawn.add(pair)

File: frontend/src/lib/flow.ts
import type {
  FlowLink,
  FlowNode,
  FlowPayload,
  FlowSide,
} from "@/lib/schemas/board";

/**
 * Reading a flow: where each box lands, where an arrow meets it, and whether
 * an arrow is long enough to carry a word.
 *
 * All of it is a pure function of the payload and the widget's shape, and none
 * of it is stored — the same division `lib/gantt.ts` makes, for the same
 * reason. The board keeps what was said; the geometry is a reading, and this is
 * where it is taken.
 *
 * Everything here is in **fractions of the widget**: `x` and `w` of its width,
 * `y` and `h` of its height, `0` its top-left and `1` its bottom-right. That is
 * `Point` in `scorsese_core::shape` read inward, and it is why a flow means the
 * same thing in a 3x2 widget and in a 16x9 one. Two fractions of different
 * things are not comparable as lengths, which is what `cells` exists to fix.
 *
 * No React and no DOM, so the arithmetic can be tested without a browser.
 */

/** A box on the widget, in fractions of it. */
export interface Box {
  x: number;
  y: number;
  w: number;
  h: number;
}

/** A place on the widget, in fractions of it. */
export interface Point {
  x: number;
  y: number;
}

/**
 * How much of its slot a box takes along the line, when nothing said where it
 * sits. The rest is the gap the arrow is drawn in, so this is also how long
 * every arrow in a fallback layout is.
 */
export const LINE_FILL = 0.68;

/** How much of the widget's short side a box takes across the line. */
export const LINE_CROSS = 0.46;

/**
 * The shortest arrow worth putting any word on, in cells.
 *
 * A cell is roughly 60px on the television this is read from, so one and a half
 * of them is about 90px — below which a label is not small, it is a smudge
 * lying across a line. Measured in cells rather than in fractions because a
 * fraction of a 3-cell widget and a fraction of a 12-cell one are different
 * lengths, and it is the length on the glass that decides.
 */
export const LABEL_FLOOR_CELLS = 1.5;

/**
 * How many cells one character of a label needs.
 *
 * `gantt.ts` asks only whether the bar is wide enough, ignoring how long the
 * name is, and it can: a bar's name is clipped by the bar. A link label has
 * nothing to clip it, so a long word on a short arrow overhangs both ends and
 * reads as a mistake in the drawing rather than in the data. Hence the length.
 */
export const LABEL_CELLS_PER_CHAR = 0.2;

/**
 * How far an S bows out of each end, as a fraction of the run between the two
 * anchors.
 *
 * A half puts each control point the familiar connector's distance out. It is
 * not a field for the same reason the bow's axis is not one — scorsese: an
 * author drawing a diagram has an opinion about which two things are joined,
 * not about the curvature of the join.
 */
export const BOW = 0.5;

/**
 * Where every box in this flow sits.
 *
 * Either they all said — in which case this is what they said, unchanged — or
 * none did, and they are drawn in one evenly spaced line along the widget's
 * longer side. The backend refuses the mixture, so there is no third case.
 *
 * The fallback is deliberately dumb: one line, uniform boxes, evenly spaced.
 * It is enough that `add_flow` with nothing but steps and arrows produces
 * something worth looking at, and it is the seam a ranked layout plugs into —
 * everything downstream of here reads boxes and knows nothing about how they
 * were chosen.
 */
export function boxes(
  nodes: FlowNode[],
  cols: number,
  rows: number,
): Map<string, Box> {
  const placed = new Map<string, Box>();
  const along = cols >= rows ? "x" : "y";
  nodes.forEach((node, at) => {
    placed.set(node.id, said(node) ?? inLine(at, nodes.length, along));
  });
  return placed;
}

/** The box a node named, or null when it named none. */
function said(node: FlowNode): Box | null {
  if (node.x === null || node.y === null || node.w === null || node.h === null)
    return null;
  return { x: node.x, y: node.y, w: node.w, h: node.h };
}

/** One box's slot in an evenly spaced line of `count` along the long side. */
function inLine(at: number, count: number, along: "x" | "y"): Box {
  const slot = 1 / count;
  const length = slot * LINE_FILL;
  const near = slot * at + (slot - length) / 2;
  const across = (1 - LINE_CROSS) / 2;
  return along === "x"
    ? { x: near, y: across, w: length, h: LINE_CROSS }
    : { x: across, y: near, w: LINE_CROSS, h: length };
}

/**
 * Where on a box a side is, as fractions of the widget.
 *
 * A rectangle and an ellipse answer this identically, which is not a shortcut:
 * the ellipse inscribed in a box touches that box at exactly the midpoints of
 * its four sides, so the anchor is the same point for both and the shape never
 * enters the arithmetic.
 */
export function anchor(box: Box, side: FlowSide): Point {
  const middle = { x: box.x + box.w / 2, y: box.y + box.h / 2 };
  switch (side) {
    case "left":
      return { x: box.x, y: middle.y };
    case "right":
      return { x: box.x + box.w, y: middle.y };
    case "top":
      return { x: middle.x, y: box.y };
    case "bottom":
      return { x: middle.x, y: box.y + box.h };
    case "center":
      return middle;
  }
}

/**
 * The pair of sides two boxes face each other across, when the link named none.
 *
 * Whichever axis their middles are further apart on, because that is the axis
 * the eye reads the relationship along — the same rule scorsese infers an S's
 * bow axis from, kept the same here so the two never disagree about which way
 * an arrow is going.
 *
 * This is the deliberate divergence from scorsese, which makes the author
 * choose. There an attached clip moves over time, so an arrow that picked its
 * own side would rearrange itself between two renders. Nothing in a flow moves.
 */
export function facing(from: Box, to: Box): [FlowSide, FlowSide] {
  const dx = to.x + to.w / 2 - (from.x + from.w / 2);
  const dy = to.y + to.h / 2 - (from.y + from.h / 2);
  if (Math.abs(dx) >= Math.abs(dy))
    return dx >= 0 ? ["right", "left"] : ["left", "right"];
  return dy >= 0 ? ["bottom", "top"] : ["top", "bottom"];
}

/** An arrow, once its geometry has been worked out. */
export interface Route {
  start: Point;
  end: Point;
  /** The two Bézier control points, or null for a straight line. */
  control: [Point, Point] | null;
  /** Which way the line is travelling as it arrives at `end`. */
  atEnd: Point;
  /** Which way it is travelling as it leaves `start`. A head there points back. */
  atStart: Point;
}

/** Which way a side faces out of its box. `center` faces nowhere. */
function normal(side: FlowSide): Point | null {
  switch (side) {
    case "left":
      return { x: -1, y: 0 };
    case "right":
      return { x: 1, y: 0 };
    case "top":
      return { x: 0, y: -1 };
    case "bottom":
      return { x: 0, y: 1 };
    case "center":
      return null;
  }
}

/**
 * Where one arrow runs, given the boxes at its two ends.
 *
 * An S leaves each end along that end's **own outward normal** rather than
 * along the dominant axis of the run. For two boxes side by side — right out of
 * one, left into the next — that is exactly scorsese's rule and draws exactly
 * scorsese's shape. It differs only where scorsese could not be asked: two
 * stacked boxes joined right-side to right-side bow out to the right and back,
 * where inferring from the run would have sent the line down out of a side it
 * leaves horizontally, which reads as a fault in the renderer.
 *
 * A `center` end has no side to leave along, so it leaves along the run — an S
 * into the middle of a box is a straight line, honestly.
 */
export function route(from: Box, to: Box, link: FlowLink): Route {
  const [a, b] = sides(from, to, link);
  const start = anchor(from, a);
  const end = anchor(to, b);
  const straight = unit(end.x - start.x, end.y - start.y) ?? { x: 1, y: 0 };
  if (link.curve === "straight")
    return { start, end, control: null, atEnd: straight, atStart: straight };

  const reach = Math.hypot(end.x - start.x, end.y - start.y) * BOW;
  const out = normal(a) ?? straight;
  const back = normal(b) ?? { x: -straight.x, y: -straight.y };
  const control: [Point, Point] = [
    { x: start.x + out.x * reach, y: start.y + out.y * reach },
    { x: end.x + back.x * reach, y: end.y + back.y * reach },
  ];
  // A cubic's tangent at each end runs from the end to the control point beside
  // it. When the bow collapses — the control point landing on its own end —
  // that vector is zero and the straight line between the ends is the honest
  // fallback.
  return {
    start,
    end,
    control,
    atEnd: unit(end.x - control[1].x, end.y - control[1].y) ?? straight,
    atStart: unit(control[0].x - start.x, control[0].y - start.y) ?? straight,
  };
}

/** The two sides this link meets: what it named, else the facing pair. */
function sides(from: Box, to: Box, link: FlowLink): [FlowSide, FlowSide] {
  const picked = facing(from, to);
  return [link.source_side ?? picked[0], link.target_side ?? picked[1]];
}

/** A direction of length one, or null when there is no direction at all. */
function unit(dx: number, dy: number): Point | null {
  const length = Math.hypot(dx, dy);
  if (!Number.isFinite(length) || length <= 0) return null;
  return { x: dx / length, y: dy / length };
}

/**
 * How long a straight run across the widget actually is, in board cells.
 *
 * The two fractions are of different things, so they cannot be squared and
 * added as they stand: a run of 0.5 across a 12-cell widget is four times the
 * run of 0.5 down a 3-cell one. Scaling each by the widget's own extent puts
 * them in the same unit first.
 */
export function cells(
  from: Point,
  to: Point,
  cols: number,
  rows: number,
): number {
  return Math.hypot((to.x - from.x) * cols, (to.y - from.y) * rows);
}

/**
 * Whether an arrow this long can carry this word.
 *
 * The move `gantt.ts` makes with `roomy()` and `clock.tsx` with the date: a
 * word half-overlapping an arrow is worse than no word, so it comes off rather
 * than being shrunk until it is unreadable anyway.
 */
export function roomy(label: string, run: number): boolean {
  return (
    run >= Math.max(LABEL_FLOOR_CELLS, label.length * LABEL_CELLS_PER_CHAR)
  );
}

/** Where a label sits on its arrow: the midpoint, nudged off the line. */
export function midpoint(run: Route): Point {
  if (run.control === null)
    return {
      x: (run.start.x + run.end.x) / 2,
      y: (run.start.y + run.end.y) / 2,
    };
  // The point halfway along a cubic, which on a bowed arrow is nowhere near
  // halfway between its ends.
  const [a, b] = run.control;
  return {
    x: (run.start.x + 3 * a.x + 3 * b.x + run.end.x) / 8,
    y: (run.start.y + 3 * a.y + 3 * b.y + run.end.y) / 8,
  };
}

/** Every arrow in a flow, already routed, with the ones that lead nowhere gone. */
export function arrows(
  payload: FlowPayload,
  placed: Map<string, Box>,
): { link: FlowLink; run: Route }[] {
  return payload.links.flatMap((link) => {
    const from = placed.get(link.source);
    const to = placed.get(link.target);
    // The backend refuses a link naming a node that is not there, so this is
    // only ever a board file written by hand. Dropping the arrow is better than
    // drawing it from nowhere.
    if (!from || !to) return [];
    return [{ link, run: route(from, to, link) }];
  });
}
